src/script/script.h

```diff
@@ -414,6 +425,66 @@ class CScriptNum {
         return result;
     }
 
+    /**
+     * Computes a + b and stores it in result. If the sum overflows or results
+     * in an invalid Script integer, return true, otherwise false.
+     */
+    static bool overflow_add_int64(int64_t a, int64_t b, int64_t &result) {
+#if HAVE_DECL___BUILTIN_SADDLL_OVERFLOW
+        static_assert(std::numeric_limits<long long int>::max() ==
+                      std::numeric_limits<int64_t>::max());
+        long long int tmp_result;
+        if (__builtin_saddll_overflow(a, b, &tmp_result)) {
+            return true;
+        }
+        result = tmp_result;
+#else
+        if (a > 0 && b > std::numeric_limits<int64_t>::max() - a) {
+            // integer overflow
+            return true;
+        }
+        if (a < 0 && b < std::numeric_limits<int64_t>::min() - a) {
+            // integer underflow
+            return true;
+        }
+        result = a + b;
+#endif
+        if (result == std::numeric_limits<int64_t>::min()) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Computes a - b and stores it in result. If the difference overflows or
+     * results in an invalid Script integer, return true, otherwise false.
+     */
+    static bool overflow_sub_int64(int64_t a, int64_t b, int64_t &result) {
+#if HAVE_DECL___BUILTIN_SSUBLL_OVERFLOW
+        static_assert(std::numeric_limits<long long int>::max() ==
+                      std::numeric_limits<int64_t>::max());
+        long long int tmp_result;
```

Comment:
Why this instead of `int64_t` ?

Reply:
Otherwise we get a compile error. Both types are not treated the same.